db/records/operations/group.py

```diff
@@ -1,123 +1,240 @@
-from sqlalchemy import select, Column, func
+from enum import Enum
+import json
+import logging
+from sqlalchemy import select, func, and_, case, literal
 
-from db.records.exceptions import BadGroupFormat, GroupFieldNotFound
-from db.records.operations.select import get_query, apply_filters
+from db.records import exceptions as rec_exc
 from db.records.utils import create_col_objects
-from db.utils import execute_query
 
+logger = logging.getLogger(__name__)
 
-def append_distinct_tuples_to_filter(distinct_tuples):
-    filters = []
-    for col, value in distinct_tuples:
-        filters.append({
-            "field": col,
-            "op": "==",
-            "value": value,
-        })
-    return filters
+MATHESAR_GROUP_METADATA = '__mathesar_group_metadata'
 
 
-def get_distinct_tuple_values(
-        column_list, engine, table=None, limit=None, offset=None, output_table=None
-):
+class GroupMode(Enum):
+    DISTINCT = 'distinct'
+    PERCENTILE = 'percentile'
+
+
+class GroupMetadataField(Enum):
+    COUNT = 'count'
+    GROUP_ID = 'group_id'
+    FIRST_VALUE = 'first_value'
+    LAST_VALUE = 'last_value'
+
+
+class GroupBy:
+    def __init__(
+            self, columns, mode=GroupMode.DISTINCT.value, num_groups=None
+    ):
+        self._columns = tuple(columns) if type(columns) != str else tuple([columns])
+        self._mode = mode
```

Comment:
Why do we need underscores here? (why not have `self.mode = mode` and kill the property?)

Reply:
This gets instantiated and passed around, and I don't want any unexpected mutation.  These should be read only after instantiation.